crates/huub/src/constraints/int_value_precede.rs

```diff
@@ -0,0 +1,1144 @@
+	/// be ordered.
+	pub(crate) vars: Vec<IntDecision>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+/// Bounds propagator for the `seq_precede_chain_int` constraint.
+pub struct IntSeqPrecedeChainBounds {
+	/// List of integer variables where first occurrences of all i>0 must be
+	/// ordered.
+	vars: Vec<IntView>,
+	/// True if initial pass is completed.
+	initialized: bool,
+	/// First possible occurrence of i.
+	first: Vec<TrailedInt>,
+	/// Last possible occurrence of i.
+	last: Vec<TrailedInt>,
+	/// Used for incremental updates of upper bounds, `first[i] = k` implies
+	/// `first_val[k] = i`.
+	first_val: Vec<TrailedInt>,
+	/// Greatest i that has to occur.
+	max_last: TrailedInt,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+/// Representation of the `value_precede_chain_int` constraint within a model.
+///
+/// This constraint enforces that the first occurrences of the elements of the
+/// given integer list in the list of integer decision variables are ordered
+/// according to the given list.
+pub struct IntValuePrecedeChain {
+	/// List of integers that need to occur in order
+	pub(crate) values: Vec<IntVal>,
+	/// List of integer decision variables where first occurrences of specified
+	/// values must be ordered.
+	pub(crate) vars: Vec<IntDecision>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+/// Value consistent propagator for the `value_precede_chain` constraint.
+pub struct IntValuePrecedeChainValue {
+	/// List of integers that need to occur in order
+	values: Vec<IntVal>,
+	/// List of integer variables where first occurrences of specified values must
+	/// be ordered.
+	vars: Vec<IntView>,
+	/// True if initial pass is completed.
+	initialized: bool,
+	/// First possible occurrence of `values[i]`.
+	first: Vec<TrailedInt>,
+	/// Last possible occurrence of `values[i]`.
+	last: Vec<TrailedInt>,
+	/// Used for incremental updates of upper bounds, `first[i] = k` implies
+	/// `first_val[k] = i`.
+	first_val: Vec<TrailedInt>,
+	/// Greatest i such that `values[i]` has to occur.
+	max_last: TrailedInt,
+	/// Minimum value in values.
+	min_val: IntVal,
+	/// Maximum value in values.
+	max_val: IntVal,
+	/// Minimum value with `min_val<min_hole<max_val` such that min_hole is not an
+	/// element of values.
+	min_hole: IntVal,
+	/// Used to iterate through the holes in values.
+	next_hole: Vec<IntVal>,
+	/// List of holes in values.
+	holes: Vec<IntVal>,
+	/// Reverse mapping of actual values to their indices in the `values` array
+	mapping: Vec<Option<usize>>,
+}
+
+impl<S: SimplificationActions> Constraint<S> for IntSeqPrecedeChain {
+	fn simplify(&mut self, actions: &mut S) -> Result<SimplificationStatus, ReformulationError> {
+		let mut ub = 0;
+		for &v in self.vars.iter() {
+			if actions.check_int_in_domain(v, ub + 1) {
+				ub += 1;
+			}
+			actions.set_int_upper_bound(v, ub)?;
+		}
+		//todo this can become more powerful if updated upper bound from previous loop is available
```

Comment:
I'm not sure whether I understand this TODO. Do you mean what the change in upper bound was, or just that it would be more powerful if this is done multiple times?